### ISSUE TITLE
require('codesnippet') from players personal code snippets book

### ISSUE THREAD

**Issue by serapath** [codecombat/codecombat#2878]
In order to improve the teaching, please add a `CodeSnippetsBook` where player can add **re-usable** CodeSnippets by writing something like:

#### [CodeSnippetsBook - PAGE [1/N]
```javascript
// "FindEasyMoney" Snippet
module.exports = function getNextCoin(SELF) {
    var coins = SELF.findByType("coin");
    for(var i=0, gold=[]; i<coins.length; i++) {
        if (coins[i].bountyGold > 1) { gold.push(coins[i]); }
    }
    return SELF.findNearest(gold);
};
```
#### [CodeSnippetsBook - PAGE [2/N]
```javascript
// "MyCleaveAttackStyle" Snippet
module.exports = function attackWithCleave(SELF) {
    for (var i=0, enemies = SELF.findEnemies(), closeEnemies=[]; i<enemies.length; i++) {
        if (SELF.distanceTo(enemies[i]) < 10) {
            closeEnemies.push(enemies[i]);
        }
    }
    if (closeEnemies.length > 2 && SELF.isReady('cleave')) {
        SELF.cleave();
    } else if (enemies) {
        SELF.attack(SELF.findNearest(enemies));
    }
};
```
#### Now in his next game mission
With this, the learner could now use his personal "best practices" to come up with solutions.  
```javascript
var attackWithCleave = require('MyCleaveAttackStyle');
var getNextCoin = require('FindEasyMoney');

loop {
    var enemy = this.findNearestEnemy();
    if (enemy) {
        attackWithCleave(this);
    } else {
        getNextCoin(this);
    }
}
```
*(=> Maybe in the future people could even browse other players snippets - just like real developers)*

**Comment by UltCombo** [comment]
Nice idea!
Perhaps this issue can be merged with #2561? They seem to be duplicates.

I personally like the coding approach using modules more than an UI, but this is up to personal taste.  Since CodeCombat is about teaching how to code, teaching a module syntax could be a very nice addition as well.

Though, I'd personally prefer to see ECMAScript Modules syntax (introduced in ECMAScript 2015/ES6) rather than CommonJS, as the former is now part of the official language specification.

**Comment by UltCombo** [comment]
Alternatively, the modules could run with the same `this` binding as the main script, so you can just augment the `this` object and forget about the import/export syntax.

E.g., instead of
```js
module.exports = function getNextCoin(SELF) {/*...*/};
````

Just use:

```js
this.getNextCoin = function () {/*...*/};
```

**Comment by serapath** [comment]
I dislike the use of `this` a lot.  
I strongly prefer to minimize the keywords i'm using.

**So instead of:**
```js
/*
 Requirements:
  - understand 'this'
  - understand 'new'
  - understand 'prototype'
  - understand 'export'
  - understand 'import from'
*/
function Mammal(name){ 
	this.name=name;
	this.offspring=[];
}
Mammal.prototype.haveABaby=function(){ 
	var newBaby=new Mammal("Baby "+this.name);
	this.offspring.push(newBaby);
	return newBaby;
};
Mammal.prototype.toString=function(){ 
	return '[Mammal "'+this.name+'"]';
};
export default Mammal;
// USAGE:
import Mammal from 'Mammal';
var timmy = new Mammal('timmy');
var anna  = new Mammal('anna');
```
**i rather write**
```js
/*
 Requirements:
  - understand 'require' (not a special keyword, just a function)
  - understand 'module.exports' (not a special keyword, just an object)
*/
function Mammal (name) {
  var data = { name: name, offspring: [] };
  return {
    haveABaby : function () { return _haveABaby(data); },
    toString  : function () { return _toString(data);  }
  };
}
function _haveABaby (data) {
  var baby = Mammal("Baby " + data['name']);
  data['offspring'].push(baby);
  return baby;
}
function _toString (data) {
  return '[Mammal "' + data['name'] + '"]';
}
module.exports = Mammal;
// USAGE:
var Mammal = require('mammal');
var timmy = Mammal('timmy');
var anna  = Mammal('anna');
```

**Comment by serapath** [comment]
Btw, i do not think that es6 `import/export` syntax will take over.
The vast majority of javascript modules, mainly on `npm` are written in `commonJS` style.
...and what is an 'official language specification' anyway. I rather follow what emerges in the community as best practice than to follow some kind of authority that claims to be the main source of truth.
I perceive es6 as a suggestion of a range of features from a certain group of people and i might adopt some of those or maybe i wont.

I need to hear some arguments why es6 features are better than what we already have and for some features i see the benefits but for import/export i fail to see that.

**Comment by UltCombo** [comment]
> I dislike the use of `this` a lot.

I personally dislike `this` as well. However, CodeCombat uses `this` since the very first level -- it is just a reference to your hero.

You don't need to understand `this`, constructors and prototypes to play CodeCombat. This is because you can manipulate `this` to be any value you'd like via [`Function#call`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call), (and also `Function#apply`, `Function#bind`), and that is what CodeCombat does IIRC.

I simply advocated using the same pattern for reusable code snippets, since I believe it would be confusing to have `this` referencing the hero in the main level editor and having `this` reference something else in module code -- e.g. copypasting code from the level editor to a module wouldn't work.

If `this` references the same object in all "modules" (let's consider the level editor code to be a module as well), then perhaps an import/export syntax would be unnecessary. Perhaps not having an import/export syntax would be confusing with regard to what gets exported and what doesn't. This needs more thought.

As for CommonJS vs ES Modules syntax, I'd rather not dive too deep into it as it may derail the thread, but I'll post my 2 cents in the next comment.

**Comment by UltCombo** [comment]
> The vast majority of javascript modules, mainly on npm are written in commonJS style.

Yep, because CommonJS is so far the only natively supported module system in Node.js, and other tools such as Browserify also adopted it -- after all, CommonJS works great for the majority of use cases and there was no better module system when these platforms/tools were created.

> Btw, i do not think that es6 import/export syntax will take over.

The ES Modules syntax has emerged from the shortcomings of the existing module systems. It solves problems such as static analysis of a module's dependency tree, thus offering better support for tooling such as IDEs and bundlers (e.g. Browserify does not handle CommonJS `require`s which depend on a runtime value).

ES Modules export bindings instead of values, which is overall better with regard to circular dependencies.

ES Modules will also be deeply ingrained in browsers in the not-so-far future (once they implement the [Loader spec](https://whatwg.github.io/loader/)), playing a key role when HTTP/2 prioritization and server pushing resources from tool-generated dependency trees become mainstream.

The Loader spec will also allow customizing the module resolution logic, how to fetch it and even transform/normalize/transpile the module code on the fly.

The committee behind ES Modules went to great lengths to ensure that ES Modules will be the definitive, universal JavaScript module system. The other module systems paved the way to ES Modules.

... But of course, as far as I know, no JavaScript engine supports ES Modules syntax out of the box yet, hence code using ES Modules syntax must still be transpiled to CommonJS/AMD/UMD for the time being.

**Comment by serapath** [comment]
Ok, that would be really go into a long discussion.
But regarding the original topic, it's not helpful to leave "this" as something magic.
Currently, no level that i played teaches how to define functions and pass in arguments when calling them and `this` is not explained at all and remains some magic thingy.

# Why?
Why would the game make kids extremely familiar with `this`, as its probably the most common thing they are going to type, but explain it not at all? They build their mental model around the stuff they learn only to get it ripped apart completely once they leave the game environment?
Wouldn't it be nice to teach them how they can "outsource" some code into functions?

It is horrible to introduce modules to them, but then make them pass in arbitrary state into every module, with no control over what is going to be manipulated on the `this` object, making the `this` object some kind of global scope thing in the context of code combat.

**Comment by UltCombo** [comment]
I agree with your points regarding `this`. I'm not sure why this direction was taken, but I can take guesses.

CodeCombat does teach how to use functions -- IIRC the first level to use functions is Vital Powers, which is around the 80~90th free-to-play level.
The default JavaScript code for that level uses `this.fn = function() {};` syntax.
So when you call `this.fn()`, the `this` inside of the function will still reference the hero, allowing you to easily call the hero's methods with the syntax the player is used to (e.g. `this.attack()`).

Bu why actually it was decided to use `this` and not just a variable name (e.g. `hero`) to reference your hero is beyond me. We can wait for @nwinter to chime in.

I agree that leaving `this` as a mysterious construct is not good. But note that there are some inherent challenges when trying to explain `this`:

- JavaScript's `this` is complicated. There are good resources available (as written by [Angul Croll](https://javascriptweblog.wordpress.com/2010/08/30/understanding-javascripts-this/) and [Tim Caswell](http://howtonode.org/what-is-this)), but they aren't quite beginner-friendly.

- The concept of "this" does not work exactly the same across different languages. Seems like this would require either language-specific levels or the same level would have to teach different concepts depending on the language (or skip over the intricacies and teach just the "basics", which may leave a lot of room for confusion).

Anyway, seeing as functions are a relatively new addition, I'd expect to see constructors/`this` classes/`self` explained by the ~130-150th level, and modules around the 180~200th. I'm just guessing though. I wonder if @nwinter has plans for these already?

**Comment by serapath** [comment]
So, if you break it down to its basics, `this` is just another argument passed to a function.
If you call a function that doesnt specify parameters, you have two "built in" methods to access whatever is explicitly or implicitly given to the function.
1. `this`
2. `arguments`

So if you specify the parameters instead, you do not need to rely on `arguments`, but you cannot specify a name for `this` and also `this is not passed as an argument like ALL OTHER ARGUMENTS.
There are multiple ways to pass in the argument for the `this` parameter into the function.
It could be through a `bind` or `call` or `apply` or setting it through `new` or invoking the function as a method of an object.

If it's your code or you know the conventions, then `this` might work fine, even though you can never be really sure. On top of that, javascript is a huge ecosystem and you need to use code written by others, where you might always end up wondering what `this` refers to know. The same is true when debugging and stepping through the debugger. You carefully have to think about and keep track of what `this` is all the time. Its as if i would use a parameter that i call `x` and use it everywhere - with no clue to what `x` actually means. This is considered very bad practice with variable names, but for some reason `this` is popular.

And like you said, another problem is, that `this` is not helpful to anyone coming from a language in which it works different.

For now, when I'm teaching, I always start out with the first line, letting them write:
```js
var hero = this;
```
So that they can then move on in a way where everything makes sense, because once you start with utilizing `functions` or custom objects with methods, you also have to explain them `apply` or `call` and all kinds of weirdness they dont need to know to be more productice...

**Comment by UltCombo** [comment]
Excellent explanations. :smiley:
I've contemplated using `var hero = this` as well, as this is a common pattern when dealing with APIs that (ab)use `this`.
